server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -309,17 +312,72 @@ public void onFailure(Exception e) {
         }
     }
 
-    private void failDecommissionedNodes(ClusterState state) {
+    private void drainNodesWithDecommissionedAttribute(DecommissionRequest decommissionRequest) {
+        decommissionController.updateMetadataWithDecommissionStatus(DecommissionStatus.DRAINING, new ActionListener<>() {
+            @Override
+            public void onResponse(DecommissionStatus status) {
+                logger.info("updated the decommission status to [{}]", status);
+                // set the weights
+                checkDecommissionAttributeAndScheduleDecommission(decommissionRequest);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.error(
+                    () -> new ParameterizedMessage(
+                        "failed to update decommission status for attribute [{}] to [{}]",
+                        decommissionRequest.getDecommissionAttribute().toString(),
+                        DecommissionStatus.DRAINING
+                    ),
+                    e
+                );
+                // since we are not able to update the status, we will clear the voting config exclusion we have set earlier
+                clearVotingConfigExclusionAndUpdateStatus(false, false);
+            }
+        });
+    }
+
+    void checkDecommissionAttributeAndScheduleDecommission(DecommissionRequest decommissionRequest) {
+        ClusterState state = clusterService.getClusterApplierService().state();
+
+        DecommissionAttributeMetadata decommissionAttributeMetadata = state.metadata().decommissionAttributeMetadata();
+        if (decommissionAttributeMetadata == null) {
+            clearVotingConfigExclusionAndUpdateStatus(false, false);
+            return;
+        }
+        assert decommissionAttributeMetadata.status().equals(DecommissionStatus.DRAINING)
+            : "Unexpected status encountered while decommissioning nodes.";
+
+        // Schedule the node decommission process after the weights are successfully set.
+        scheduleNodesDecommissionOnTimeout(state, decommissionRequest.getDrainingTimeout());
+    }
+
+    void scheduleNodesDecommissionOnTimeout(ClusterState state, TimeValue timeoutForNodeDraining) {
         // this method ensures no matter what, we always exit from this function after clearing the voting config exclusion
         DecommissionAttributeMetadata decommissionAttributeMetadata = state.metadata().decommissionAttributeMetadata();
         DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
+
+        Set<DiscoveryNode> decommissionedNodes = filterNodesWithDecommissionAttribute(state, decommissionAttribute, false);
+        // Wait for timeout to happen. Log the active connection before decommissioning of nodes.
+        transportService.getThreadPool().schedule(() -> {
```

Comment:
Check if the delay is zero if it is it doesn't make sense to fork and we can continue in the same thread?